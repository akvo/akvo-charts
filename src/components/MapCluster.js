import React, { forwardRef } from 'react';
import MarkerClusterGroup from './Map/MarkerClusterGroup';
import Marker from './Map/Marker';
import { default as Container } from './Map/View';

const CLUSTER_TYPE = {
  default: 0,
  circle: 1
};

const clusterCircleIcon = (
  cluster,
  data = [],
  groupKey = 'name',
  props = {}
) => {
  const groupValues = Object.values(
    data?.reduce((acc, item) => {
      const groupName = item?.[groupKey];
      const color = item?.color;
      if (!acc[groupName]) {
        acc[groupName] = { value: groupName, color, count: 0 };
      }
      acc[groupName].count++;

      return acc;
    }, {})
  );
  const totalValue = groupValues.reduce((s, { count }) => s + count, 0);
  const radius = 40;
  const circleLength = Math.PI * (radius * 2);
  let spaceLeft = circleLength;
  return {
    html: `<svg width="100%" height="100%" viewBox="0 0 100 100"> <circle cx="50" cy="50" r="40" fill="#ffffffad"/>
          ${groupValues
            .map((item, index) => {
              const v = index === 0 ? circleLength : spaceLeft;
              spaceLeft -= (item.count / totalValue) * circleLength;
              return `
                <circle cx="50" cy="50" r="40" fill="transparent" stroke-width="15" stroke="${
                  item.color ? item.color : 'red'
                }" stroke-dasharray="${v} ${circleLength}" />`;
            })
            .join(
              ''
            )} <text x="50%" y="50%" fill="black" text-anchor="middle" dy=".3em" font-size="18px">${cluster.getChildCount()}</text></svg>`,
    ...props
  };
};

const MapCluster = (
  {
    data,
    markerIcon = {
      className: 'custom-marker',
      iconSize: [32, 32],
      html: true
    },
    clusterIcon = {
      className: `custom-marker-cluster`,
      iconSize: 60
    },
    groupKey = 'name',
    type = 'default',
    renderPopup = null,
    ...config
  },
  ref
) => {
  const clusterTypes = {
    [CLUSTER_TYPE.default]: null,
    [CLUSTER_TYPE.circle]: (cluster) =>
      clusterCircleIcon(cluster, data, groupKey, clusterIcon)
  };

  const iconCreateFn =
    clusterTypes?.[CLUSTER_TYPE?.[type]] || clusterTypes.default;

  return (
    <Container
      ref={ref}
      {...config}
    >
      <MarkerClusterGroup iconCreateFn={iconCreateFn}>
        {data
          ?.filter((d) => d?.point)
          ?.map((d, dx) => (
            <Marker
              latlng={d?.point}
              key={dx}
              icon={{
                ...markerIcon,
                html: markerIcon?.html
                  ? `<span style="background-color:${d?.color}; border:2px solid #fff;"/>`
                  : null
              }}
            >
              {typeof renderPopup === 'function' ? (
                renderPopup(d)
              ) : (
                <React.Fragment>{d?.label}</React.Fragment>
              )}
            </Marker>
          ))}
      </MarkerClusterGroup>
    </Container>
  );
};

export default forwardRef(MapCluster);
