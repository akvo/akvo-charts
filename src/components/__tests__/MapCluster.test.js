import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import MapCluster from '../MapCluster';

describe('MapCluster chart', () => {
  test('renders MapCluster correctly', async () => {
    const props = {
      tile: {
        url: 'https://tile.openstreetmap.org/{z}/{x}/{y}.png',
        maxZoom: 19,
        attribution: '© OpenStreetMap'
      },
      data: [
        {
          point: [39.61, -105.02],
          name: 'Basic',
          label: 'This is Littleton, CO.',
          color: 'blue'
        },
        {
          point: [39.73, -104.8],
          name: 'Limited',
          label: 'This is Aurora, CO.',
          color: 'yellow'
        }
      ],
      config: {
        center: [39.73, -104.99],
        zoom: 10,
        height: '100vh',
        width: '100%'
      }
    };
    let instance = null;
    render(
      <MapCluster
        {...props}
        ref={(el) => {
          instance = el;
        }}
      />
    );

    await waitFor(() => {
      const mapContainer = screen.getByTestId('map-view');
      expect(mapContainer).toBeInTheDocument();

      expect(instance.getMap().getPixelOrigin()).toEqual({
        x: 54621,
        y: 99498
      });
      expect(instance.getMap().getMaxZoom()).toEqual(19);
    });
  });

  test('matches MapCluster snapshot', async () => {
    const props = {
      tile: {
        url: 'https://tile.openstreetmap.org/{z}/{x}/{y}.png',
        maxZoom: 19,
        attribution: '© OpenStreetMap'
      },
      layer: {
        source: 'window.topoData'
      },
      data: [
        {
          point: [39.61, -105.02],
          name: 'Basic',
          label: 'This is Littleton, CO.',
          color: 'blue'
        },
        {
          point: [39.73, -104.8],
          name: 'Limited',
          label: 'This is Aurora, CO.',
          color: 'yellow'
        }
      ],
      config: {
        center: [39.73, -104.99],
        zoom: 10,
        height: '100vh',
        width: '100%'
      }
    };

    let instance = null;
    const { container } = render(
      <MapCluster
        {...props}
        ref={(el) => {
          instance = el;
        }}
      />
    );
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});
