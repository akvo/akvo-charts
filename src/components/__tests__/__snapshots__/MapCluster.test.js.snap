// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapCluster chart matches MapCluster snapshot 1`] = `
<div>
  <div
    class="leaflet-container leaflet-touch leaflet-grab leaflet-touch-drag leaflet-touch-zoom"
    data-testid="map-view"
    style="height: 100vh; width: 100%; position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      >
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 19; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              src="https://tile.openstreetmap.org/10/213/388.png"
              style="width: 256px; height: 256px; left: -93px; top: -170px;"
            />
          </div>
        </div>
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 19; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              src="https://tile.openstreetmap.org/10/213/388.png"
              style="width: 256px; height: 256px; left: -93px; top: -170px;"
            />
          </div>
        </div>
      </div>
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-disabled="false"
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            <span
              aria-hidden="true"
            >
              +
            </span>
          </a>
          <a
            aria-disabled="false"
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            <span
              aria-hidden="true"
            >
              −
            </span>
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JavaScript library for interactive maps"
          >
            Leaflet
          </a>
           
          <span
            aria-hidden="true"
          >
            |
          </span>
           © OpenStreetMap
        </div>
      </div>
    </div>
  </div>
</div>
`;
